src/ripple/app/ledger/TransactionMaster.h

```diff
@@ -39,7 +40,9 @@ class TransactionMaster
     TransactionMaster& operator= (TransactionMaster const&) = delete;
 
     std::shared_ptr<Transaction>
-    fetch (uint256 const& , bool checkDisk);
+    fetch (uint256 const& , bool checkDisk,
+               ClosedInterval<uint32_t> const& range = {},
+                   bool* searchedAll = nullptr);
```

Comment:
What do you think of a new overload, something like:
```
std::variant<Transaction::pointer, bool> load(uint256 const&, bool, ClosedInterval const&)
```
Rational: I want to make clear that this function returns either the transaction or a bool saying if the entire range is searched. If it returns the transaction, I want to skip the `searchedAll` calculation, and I don't want callers to think that it's valid value.

Reply:
I like this approach.

Reply:
I considered doing something similar with `boost::optional`, but the only drawback to this approach is that now we have to rely on the value of the `ClosedInterval` parameter to determine whether or not the user intended to perform the range search. We can assume that a default range of `[0,0]` means that the user didn't supply a range, but this means that in some imagined scenario where the user actually wants to check ledger 0, this won't be possible. But if we don't really care about that, then I'll just use this approach as is. The other option is to keep the variant return, but allow a bool parameter but then this is very similar to my original approach.

Reply:
The way this interface works is:
1) When we don't pass an interval, the `searchedAll` parameter isn't relevant. 
2) When we find the transaction, the `searchedAll` parameter isn't relevant. 
3) When we don't find the transaction, and we specify an interval, when we need to set `searchedAll`. 

I'm proposing that we keep the old function, and add an additional overload. I.e.:
```
// overload 1
std::shared_ptr<Transaction> 
fetch (uint256 const& , bool checkDisk);

// overload 2
std::variant<Transaction::pointer, bool> 
load(uint256 const&, bool checkDisk, ClosedInterval const&)
```
I think these two overload capture those semantics better than the single function with the default interval parameter and `searchedAll` out parameter. 

Note: My biggest complaint about the above interface is `std::variant` is a pain to use in C++, but in this case I prefer it to the alternatives. An overload returning `std::pair<shared_ptr, bool>` would be my second choice. I would avoid `tribool` or optional<bool>` as too error prone.

If we need to distinguish if the user passes a range or not, we could use an `optional<ClosedInterval>` for that. However, I prefer the two overloads to that.

Reply:
Is overload 1 supposed to be load?

Reply:
@undertome Oops. That was a typo. Both should be `fetch`.

Reply:
Understood - I’ll have think of a good way to distinguish wether or not the range should be used when you get to the load method though.

Reply:
Overload `load` as well.

Reply:
Right, but my concern was avoiding code duplication. So I went with overload stubs to mitigate that.

Reply:
Done.

Reply:
Is std::variant used elsewhere in the code? It seems to be causing build errors with clang...

Reply:
clang 5/6 and xcode 9 specifically seem to fail. No, I don't think we've used `<variant>` so far and it's looking like we'd need to trim our supported compilers or make accommodations for them (fallback to boost perhaps?)

Reply:
I'm ok with requiring xcode 10.  Requiring latest tools on platforms we don't officially support doesn't seem that draconian to me.

Reply:
Great, that solves the xcode failure - what should we do about clang 5/6?

Reply:
I'd use `boost::variant`, as @mellery451 suggests